File: DSA/RATNA_KHANNA_590013141/DAY_27_QUESTION_1.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node* prev;
    struct node* next;
};

struct node* createNode(int data) {
    struct node* newNode = (struct node*)malloc(sizeof(struct node));
    if (newNode == NULL) {
        printf("Memory allocation failed\n");
        exit(1);
    }
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}

struct node* insertEnd(struct node* head, int data) {
    struct node* newNode = createNode(data);
    if (head == NULL) {
        return newNode;
    }

    struct node* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }

    temp->next = newNode;
    newNode->prev = temp;

    return head;
}

struct node* deleteAtPos(struct node* head, int pos) {
    if (head == NULL || pos <= 0) {
        return head;
    }

    struct node* temp = head;
    int i = 1;

    while (temp != NULL && i < pos) {
        temp = temp->next;
        i++;
    }

    if (temp == NULL) {
        return head;
    }

    if (temp->prev != NULL) {
        temp->prev->next = temp->next;
    } else {
        head = temp->next;
    }

    if (temp->next != NULL) {
        temp->next->prev = temp->prev;
    }

    free(temp);
    return head;
}

void printList(struct node* head) {
    struct node* temp = head;
    while (temp != NULL) {
        printf("%d", temp->data);
        if (temp->next != NULL) {
            printf(" <--> ");
        }
        temp = temp->next;
    }
    printf("\n");
}

int main() {
    struct node* head = NULL;
    int n, val, pos;

    printf("Enter number of elements: ");
    scanf("%d", &n);

    printf("Enter %d elements:\n", n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &val);
        head = insertEnd(head, val);
    }

    printf("Enter position to delete (1-based): ");
    scanf("%d", &pos);

    head = deleteAtPos(head, pos);

    printf("Final list:\n");
    printList(head);

    return 0;
}

File: DSA/VenikaPanwar_590013755/Day_27_Question_1.java
import java.util.*;

public class Day_27_Question_1 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        DoublyNode head = createDoublyList(sc);

        System.out.print("Enter position to delete (1-based): ");
        int pos = sc.nextInt();

        head = deleteAtPosition(head, pos);
        printDoublyList(head);
    }

    static DoublyNode deleteAtPosition(DoublyNode head, int x) {
        if (head == null) return null;

        if (x == 1) {
            head = head.next;
            if (head != null) head.prev = null;
            return head;
        }

        DoublyNode current = head;
        for (int i = 1; i < x && current != null; i++) {
            current = current.next;
        }

        if (current == null) return head;

        if (current.prev != null)
            current.prev.next = current.next;

        if (current.next != null)
            current.next.prev = current.prev;

        return head;
    }

    static DoublyNode createDoublyList(Scanner sc) {
        System.out.print("Enter number of elements: ");
        int n = sc.nextInt();
        DoublyNode head = null, tail = null;
        System.out.println("Enter elements:");
        for (int i = 0; i < n; i++) {
            DoublyNode node = new DoublyNode(sc.nextInt());
            if (head == null) head = tail = node;
            else {
                tail.next = node;
                node.prev = tail;
                tail = node;
            }
        }
        return head;
    }

    static void printDoublyList(DoublyNode head) {
        while (head != null) {
            System.out.print(head.data);
            if (head.next != null) System.out.print(" <-> ");
            head = head.next;
        }
        System.out.println();
    }
}
